export.go

```diff
@@ -6,6 +6,15 @@ import (
 	"fmt"
 )
 
+type OrderType int
+
+// OrderTraverse is the type of traversal order to use when exporting and importing.
+// PreOrder is needed for the new node-key refactoring. The default is PostOrder.
```

Comment:
I thought new node-key refactoring works with both orderings?

Reply:
no, the pre-order is needed for node-key refactoring

Reply:
slightly confused now. This pr adds support for both pre and post order, but node key refactor require pre-order. If a node exports using post-order then we cant import it into the new version correct?

Reply:
yes, that's why provides both orders

Reply:
even it is the original version, we can request a pre-order snapshot, then import it into the new version

Reply:
I don't think this flow works with the sdk. Users will snapshot what they know, post-order, then when a node using this new version requests chunks, im not sure it can ask for pre or post order, it just gets what the other node has. I believe this flow has a negotiation between peers on which snapshot?

Reply:
If we can add the param which denote pre-order or post-order in Snapshot of cosmos-sdk, we don't need negotiation.

- Firstly, we create the node with original version and sync the original snapshot (post-order) and save (pre-order) snapshot (it is possible if we merge this PR, since we provide both-way import/export)
- Every nodes which want new version start sync from the created node, and get (pre-order) snapshot and they can rebuild the new version iavl tree

Reply:
Currently, in the state-sync protocol, node can fetch the snapshot chunks from multiple nodes, they are expected to be consistent and deterministic, there's no way for client node to filter different snapshot formats.

Reply:
but we can point the source nodes in the config, right? In the second step, we can set the source nodes as the new created nodes which provide pre-order snapshot

Reply:
so I think this makes sense, cause on start of the node, it will export using pre-order and ingest the same snapshot to recreate state and move the chain ahead. The question becomes how to support historical queries to the older state or if that is a needed feature? 

This would mean that a chain that upgraded to this version would share updated snapshots. The only question comes from a node using an older version of iavl which I don't think we will support in realtime, but even the old way can be done pre-order correct?

Reply:
I think the reason post-order is chosen in the first place is the node hash is updated in a post-order way, you need to update the children first to update the parent node, will pre-order import need more temporary memory?

Reply:
yeah, literally post-order more makes sense, both way requires stack to keep the current path, you are right pre-order will require 2 times memory, but the stack length is at most the height of the tree, it is so trivial

Reply:
@tac0turtle , the old version can use any order, the new version should use pre-order.

Reply:
> but the stack length is at most the height of the tree, it is so trivial

yeah, that should be trivial then, if this is indeed necessary, I think the chains can do a coordinated upgrade in advance to switch to pre-order snapshots, before doing node key format migration.

Reply:
> yeah, that should be trivial then, if this is indeed necessary, I think the chains can do a coordinated upgrade in advance to switch to pre-order snapshots, before doing node key format migration.

I think this is what we should discuss, tbh I have no clear idea which way is more efficient